finat/physically_mapped.py

```diff
@@ -268,15 +268,13 @@ def basis_evaluation(self, order, ps, entity=None, coordinate_mapping=None):
         assert coordinate_mapping is not None
 
         M = self.basis_transformation(coordinate_mapping)
-        M, = gem.optimise.constant_fold_zero((M,))
+        csr = [[j for j in range(M.shape[1]) if not isinstance(M.array[i, j], gem.Zero)]
+               for i in range(M.shape[0])]
 
         def matvec(table):
-            table, = gem.optimise.constant_fold_zero((table,))
-            i, j = gem.indices(2)
-            value_indices = self.get_value_indices()
-            table = gem.Indexed(table, (j, ) + value_indices)
-            val = gem.ComponentTensor(gem.IndexSum(M[i, j]*table, (j,)), (i,) + value_indices)
-            # Eliminate zeros
+            table = [gem.partial_indexed(table, (j,)) for j in range(M.shape[1])]
+            expressions = [sum(M.array[i, j] * table[j] for j in js) for i, js in enumerate(csr)]
```

Comment:
If these are numpy arrays it is more efficient to use slicing and such instead of explicit loops. Does that apply here? Is this line ever expensive?

Reply:
I tried turning the table into a numpy array, and grouping the blocks that shared the same rows to use @ instead of `sum(a[i,  j] * b[j] for j in js)`, but that resulted in slower compilation time than the what I currently have here.  The number of terms in the sums is expected to be very small, around 3.

Reply:
Oh that's definitely fine then. Thanks!